apiserver/facades/client/modelupgrader/upgrader.go

```diff
@@ -110,45 +120,149 @@ func (m *ModelUpgraderAPI) AbortModelUpgrade(arg params.ModelParam) error {
+	defer func() {
+		result.ChosenVersion = targetVersion
+	}()
+
 	modelTag, err := names.ParseModelTag(arg.ModelTag)
 	if err != nil {
-		return errors.Trace(err)
+		return result, errors.Trace(err)
 	}
 	if canWrite, err := m.hasWriteAccess(modelTag); err != nil {
-		return errors.Trace(err)
+		return result, errors.Trace(err)
 	} else if !canWrite && !m.isAdmin {
-		return apiservererrors.ErrPerm
+		return result, apiservererrors.ErrPerm
 	}
 
 	if err := m.check.ChangeAllowed(); err != nil {
-		return errors.Trace(err)
+		return result, errors.Trace(err)
+	}
+
+	// We now need to access the state pool for that given model.
+	st, err := m.statePool.Get(modelTag.Id())
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	defer st.Release()
+
+	model, err := st.Model()
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+
+	agentVersion, err := model.AgentVersion()
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	targetVersion, err = m.decideVersion(
+		targetVersion, agentVersion, arg.AgentStream, st, model,
+	)
+	if errors.Is(errors.Cause(err), errors.NotFound) || errors.Is(errors.Cause(err), errors.AlreadyExists) {
+		result.Error = apiservererrors.ServerError(err)
+		return result, nil
+	}
+
+	if err != nil {
+		return result, errors.Trace(err)
 	}
 
 	// Before changing the agent version to trigger an upgrade or downgrade,
 	// we'll do a very basic check to ensure the environment is accessible.
 	envOrBroker, err := m.newEnviron()
 	if err != nil {
-		return errors.Trace(err)
+		return result, errors.Trace(err)
 	}
-	if err := environs.CheckProviderAPI(envOrBroker, m.callContext); err != nil {
-		return errors.Trace(err)
+	if err := preCheckEnvironForUpgradeModel(
+		m.callContext, envOrBroker, model, agentVersion, targetVersion,
+	); err != nil {
+		return result, errors.Trace(err)
 	}
-	if err := m.validateModelUpgrade(false, modelTag, arg.ToVersion); err != nil {
-		return errors.Trace(err)
+
+	if err := m.validateModelUpgrade(false, modelTag, targetVersion, st, model); err != nil {
+		return result, errors.Trace(err)
 	}
 	if arg.DryRun {
-		return nil
+		return result, nil
 	}
-	st, err := m.statePool.Get(modelTag.Id())
-	if err != nil {
+
+	var agentStream *string
+	if arg.AgentStream != "" {
+		agentStream = &arg.AgentStream
+	}
+	if err := st.SetModelAgentVersion(targetVersion, agentStream, arg.IgnoreAgentVersions); err != nil {
+		return result, errors.Trace(err)
+	}
+	return result, nil
+}
+
+func preCheckEnvironForUpgradeModel(
+	ctx context.ProviderCallContext, env environs.BootstrapEnviron,
+	model Model, agentVersion, targetVersion version.Number,
+) error {
+	if err := environs.CheckProviderAPI(env, ctx); err != nil {
 		return errors.Trace(err)
 	}
-	defer st.Release()
-	return st.SetModelAgentVersion(arg.ToVersion, &arg.AgentStream, arg.IgnoreAgentVersions)
+
+	if model.Name() != bootstrap.ControllerModelName || model.Owner().Name() != environs.AdminUser {
+		return nil
+	}
```

Comment:
This isn't correct - we can't just check for "admin".
But anyway, we've already done the auth check earlier in UpgradeModel(), so this isn't needed.